base/abstractdict.jl

```diff
@@ -486,6 +486,64 @@ function hash(a::AbstractDict, h::UInt)
     hash(hv, h)
 end
 
+"""
+    modify!(f, d::AbstractDict{K, V}, key)
+
+Lookup and then update, insert or delete in one go without re-computing the hash.
+
+`f` is a callable object that must take a single `Union{Some{V}, Nothing}` argument and return
+a `Union{T, Some{T}, Nothing}` value, where `T` is a type [`convert`](@ref)-able to the value type
+`V`.  The value `Some(d[key])` is passed to `f` if `haskey(d, key)`; otherwise `nothing`
+is passed.  If `f` returns `nothing`, corresponding entry in the dictionary `d`  is removed.
+If `f` returns non-`nothing` value `x`, `key => something(x)` is inserted or updated in `d`
+(equivalent to `d[key] = something(x)` but more efficient).
+
+`modify!` returns whatever `f` returns as-is.
+
+# Examples
+```jldoctest
+julia> dict = Dict("a" => 1);
+
+julia> modify!(dict, "a") do val
+           Some(val === nothing ? 1 : something(val) + 1)
+       end
+Some(2)
```

Comment:
I wonder if we should make this result `Some(1) => Some(2)`, so you get back both old and new, instead of just the new value (similar to `replacefield!`).

Reply:
This is now implemented in 28900adceb1c95a65606a07f62a178e2b16f3092.